toc/rfc/rfc-draft-add-paketo-stacks.md

```diff
@@ -0,0 +1,133 @@
+and [Static (experimental)](https://github.com/paketo-buildpacks/jammy-static-stack)
+. The Paketo Full stack is roughly equivalent to cflinuxfs4, whereas the Base,
+Tiny, and Static stacks include progressively fewer Ubuntu packages.
+
+#### Separate Build and Run Images
+
+CF Deployment uses the same rootFS for building and running applications. This
+means that packages that are only needed for building applications (e.g.
+compilers like `gcc`) are also present in running app containers. The Paketo
+stacks instead [come with both Build and Run images](https://paketo.io/docs/concepts/stacks/#what-is-a-stack)
+, where build-time Ubuntu packages are only included in the Build image.
+
+#### Conclusion
+
+Accounting for both of these factors, there are two "dimensions" to Paketo
+stacks that affect the number of Ubuntu packages included in app containers:
+weight and build/run. CF Deployment could one day offer all of these options,
+so Cloud Foundry users can select the stack that matches their app's Ubuntu
+package needs and security posture.
+
+### Goals
+
+1. Align CF Deployment with the main focus of the Buildpacks & Stacks team
+1. Bring new capabilities to CF Deployment users, including:
+    1. Lightweight Stacks
+    1. Separate Build and Run Images
+1. Increase cohesion and app portability between CF Deployment and Korifi, via
+   shared Paketo stacks
+1. Increase user base for the Paketo stacks
+1. Open the door for eventual deprecation of the cflinuxfs* stacks, reducing
+   maintenance costs (cflinuxfs* deprecation is NOT included in this RFC)
+
+### High-Level Implementation
+
+Instead of using cflinuxfs4, apps using the new CNBs will use the
+[Paketo Jammy Full stack](https://github.com/paketo-buildpacks/jammy-full-stack)
+. The Jammy Full stack will include two rootFSs: build and run. The Paketo
+Jammy Full stack is very similar to cflinuxfs4, which will reduce the
+difficulty of migrating existing Cloud Foundry apps to the Paketo stacks and
+buildpacks. This will also lay the groundwork for eventually supporting the
+remaining Paketo Jammy stacks, which will offer better security posture for
+interested Cloud Foundry users.
+
+#### App Developer Experience
+
+The Paketo Full stack will be modeled as a single Cloud Foundry stack with two
+associated rootFSs. The details of selecting the rootFS for build and run
+containers will be handled by CAPI, depending on whether it is staging an app
+or running a process/task. This means that app developers will not need to be
+aware that separate build and run rootFSs are being used, they only need to
+push their app with the new stack:
+
+```sh
+cf push my-app --stack=paketo_jammy_full
+```
+
+#### Initial Scope
+
+1. Produce a tarball of the Paketo Full stack (instead of an OCI image)
+1. Create a BOSH release for the Paketo Full stack
+1. Update CNB shims to use the Paketo Full stack instead of cflinuxfs4
+1. Add CF support for separate build and run rootFSs
```

Comment:
I have two points here:

1. My understanding from the [RFC for the first phase](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0017-add-cnbs.md) was that it aims to introduce CNB experience in CF without requiring any changes in the CF platform itself. Is that correct or this RFC will be required to try the shimmed CNBs in a CF environment? 
2. I really like the idea to introduce support for build and run rootFSs in CF. This could be beneficial also for the current V2 buildpacks.  My question is around the plan to create a tarball of the Paketo Full stack and introduce a rootFSs for the CNB use case. If I understand that correctly the approach will deviate a lot from the [CNB Platform Interface Specification](https://github.com/buildpacks/spec/blob/main/platform.md#stacks) and I think it will make an eventual support of OCI based app builds at a later point of time harder to achieve. My suggestion is that we should evaluate at this point what does it mean to go with a more CNB closed lifecycle approach vs. V2 based one. We should put out of scope the registry part in the [CNB specification](https://github.com/buildpacks/spec/blob/main/platform.md#exporter) here but at least the rest should be taken in consideration. 

I'm really interested to get other opinions on this.

Reply:
> My understanding from the [RFC for the first phase](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0017-add-cnbs.md) was that it aims to introduce CNB experience in CF without requiring any changes in the CF platform itself. Is that correct or this RFC will be required to try the shimmed CNBs in a CF environment?

Without this RFC, the plan is to release the CNB shims on the cflinuxfs4 stack. If this RFC is adopted, then that plan may be modified.

Buildpacks can be compatible with multiple stacks, so it is an option to release the CNB shims for both cflinuxfs4 and the Paketo stacks, especially since they are both based on Ubuntu 22.04 Jammy. It's mostly just a question of whether it will be worth the effort to release on both stacks.

I'm happy to amend this RFC based on what reviewers think the best option is for CNB stack compatibility.

Reply:
> I really like the idea to introduce support for build and run rootFSs in CF. This could be beneficial also for the current V2 buildpacks.

It's certainly a possibility that we could extend this to the v2 buildpacks. I don't think we currently have bandwidth to do that, since most of the focus of the buildpacks team is on Paketo and CF Deployment Paketo integration right now. It would also be tricky to roll out for cflinuxfs4, since it would be a breaking change for the stack to lose a bunch of OS packages at runtime.

Reply:
> My question is around the plan to create a tarball of the Paketo Full stack and introduce a rootFSs for the CNB use case. If I understand that correctly the approach will deviate a lot from the [CNB Platform Interface Specification](https://github.com/buildpacks/spec/blob/main/platform.md#stacks) and I think it will make an eventual support of OCI based app builds at a later point of time harder to achieve.

From our experiments, it's fairly easy to translate between the Paketo OCI images and CF rootFS tgzs, so this was intended to be a low-effort step to incremental Paketo adoption in CF Deployment.

From my perspective, this RFC steps us slightly closer to CNB Platform adoption by:
1. Using the Paketo stacks
2. Enabling differences between the build and run environments

That said, I'm by no means an expert on the CNB Platform spec. I'd love to hear more details about your concerns on how this might make future steps more difficult.

cc @robdimsdale

Reply:
Paketo stacks are OCI images, which is to say they are essentially a filesystem with metadata. Cloud Foundry stacks are just filesystems in a tgz format. Therefore the simplest way to get a Paketo stack into Cloud Foundry is to get a tgz of the stack image's filesystem.

It's true that this approach discards the OCI image metadata associated with the filesystem, but that doesn't prevent the filesystem from working with Cloud Foundry today. Additionally, I don't see how adopting tarballs of the filesystems as a first step precludes future support for the OCI image stacks directly.

Am I missing something here?

Reply:
> - I’m not so familiar with the SBOM generation but does the generated SBOM describe the stack used for the app?
>   - If yes, then a future update of the stack via the CF platform will invalidate the generated SBOM of an app. We should decide how to deal with this depending on our decisions.

This is an excellent point, I'm also curious about how the SBOM generation part works with the build-and-run stacks. If the SBOM is generated on the build stack it will show dependencies that might not be available on the run stack.

Reply:
For the Paketo stacks, we provide SBOMs for both build and run images (e.g. [build receipt](https://github.com/paketo-buildpacks/jammy-base-stack/releases/download/v0.1.68/jammy-base-stack-0.1.68-run-receipt.cyclonedx.json), [run receipt](https://github.com/paketo-buildpacks/jammy-base-stack/releases/download/v0.1.68/jammy-base-stack-0.1.68-run-receipt.cyclonedx.json)). Similar we provide SBOMs for each of the buildpacks.

Merging the two is a responsibility of the Platform. E.g. `pack`/`kpack` are responsible for this. In a CF world, the tool that provides the V2<->V3 API wrapper would be acting as the Platform, and so would have to collate and expose the SBOMs from across the image layers, and present that to the CF API.

Reply:
In the current world where stacks are not included in droplets, we would only want to include buildpack SBOMs, right? The stack can change for the application without requiring a restage.